[Exchange Online Message Trace] Add Additional Look-back Time & Fix Cursor Value

BenB196 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

Enhancement & Bug

## Proposed commit message

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

This PR makes two changes to the Exchange Online Message Trace integration:

~1. Performs a refactor of `httpjson.yml.hbs` file to move the `request.*` & `response.*` parts together, or inversely moving all the fields/tags stuff towards the end of the file. (making it easier to follow from my perspective)~
1. Fixes an issue with how the value for `last_execution_datetime` cursor was being set.
    - Previously, the cursor value was using `now`, the issue with this is that the cursor sets it value once ["all events ... are published"](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-httpjson.html#cursor). This meant that if a cycle took five minutes to complete, there could be a five minute gap in time/data between the last iteration's `EndDate` and the next iteration's `StartDate`. This appears to have always been an "issue" but became more noticeable with #8550 where it changed the default interval from `1m` to `1h`, making it take far longer to process an iteration and making the gap in data more noticable.
    - ~To "fix" this issue, rather than using `now`, we take the `.last_event.EndDate` as the value. This should ensure that the next iteration always starts where the last one ended.~
    - ~Side Note: This also fixes another issue, for if Microsoft ever has an issue publishing Message Traces. Before this change, the integration would continue on as if there was no problem and leave a potentially large gap in data. Now the integration will hold at the last published event and pick up once Microsoft fixes the event publishing.~
2. Adds the ability to configure an additional look-back time to ensure data is properly collected.
    - See: https://github.com/elastic/integrations/pull/8717#issuecomment-1859322741

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ] I ran `elastic-package.exe test` and the tests passed, however, I'm honestly not too familiar with the httpjson input. Would someone be able to take a look and confirm that this does indeed look correct, or if there is any additional testing that can be done? Looking at how some of the other cursors do it; https://github.com/search?q=repo%3Aelastic%2Fintegrations+%22cursor%3A%22&type=code this does appear to be "correct".
- [ ] There is a potential "bug" here in that if `last_execution_datetime` ever gets too far out of date (older than Microsoft's Message Trace retention duration - 90 days), then the integration would be "stuck" in this state until recreated. This seems like it would be a similar issue with a number of other integrations, so I opted not to try and solve it here as I'm not sure how "real" of a "bug" this actually is.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

- I ran `elastic-package.exe test` and everything passed.
- I validated that the integration yaml updated correctly in Kibana/Fleet.

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Closes #8674
    - Found the cursor issue while investigating this issue.
- Relates #8550

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

(Interval for the screenshots set to `5m` for easier visibility into the change)

Current/`main`/`1.17.x`:

![image](https://github.com/elastic/integrations/assets/8277432/5fee8fd9-6615-4769-8232-bebf846bd4fa)

This PR:

![image](https://github.com/elastic/integrations/assets/8277432/9b43d6b7-2ab9-4e7f-9ac7-5fb034da0eca)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8717/detail/PR-8717/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8717/detail/PR-8717/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8717/detail/PR-8717/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8717/detail/PR-8717/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-12-13T01:44:06.637Z&rangeTo=2023-12-13T02:04:06.637Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=65d97cf246b4923867f85f7b0451dc07&transactionId=9dcd2d76dac41b17)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-12-13T01:54:06.637+0000


* Duration: 3 min 23 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8717/detail/PR-8717/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Load a resource file from a library`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-8717/runs/1/steps/81/log/?start=0">here</a></li>
<li>Description: <code>approval-list/elastic/integrations.yml</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-8717/runs/1/steps/89/log/?start=0">here</a></li>
<li>Description: <code>githubApiCall: The REST API call https://api.github.com/orgs/elastic/members/BenB196 return the message : java.lang.Exception: httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/BenB196 : httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/BenB196 : Code: 404Error: {"message":"User does not exist or is not a member of the organization","documentation_url":"https://docs.github.com/rest/orgs/members#check-organization-membership-for-a-user"}</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

efd6 commented on packages/microsoft_exchange_online_message_trace/data_stream/log/agent/stream/httpjson.yml.hbs:
```diff
@@ -36,23 +36,6 @@ request.transforms:
   - set:
       target: url.params.$skiptoken
       value: 0
-fields_under_root: true
```
Let's not do this in this PR.

BenB196 replied:
Reverted

efd6 commented on packages/microsoft_exchange_online_message_trace/data_stream/log/agent/stream/httpjson.yml.hbs:
```diff
@@ -70,7 +53,25 @@ response.pagination:
 cursor:
   last_execution_datetime:
-    value: '[[(formatDate (now) "RFC3339")]]'
+    value: '[[.last_event.EndDate]]'
```
Do we know that the events in the value array are ordered ascending? This will not work if they are not. If we can't guarantee it, we should conditionally use the later of `.last_event.EndDate` and `.cursor.last_execution_datetime`

ShourieG replied:
From my own time playing around with the message trace api, the events  are ordered in ascending manner. @BenB196 could you confirm this since you have a larger dataset available to check. If true, your fix should work fine and we could proceed with this PR after the other issues are resolved. Also if possible a system test to simulate this behaviour and test this implementation would be awesome.

BenB196 replied:
The API does support an [`orderby`](https://learn.microsoft.com/en-us/previous-versions/office/developer/o365-enterprise-developers/jj984335(v=office.15)#fields) parameter, but its not clear what the default order is. I'll investigate a bit more and determine whether this needs to be adjusted.

> Also if possible a system test to simulate this behavior and test this implementation would be awesome.

I'll see if I can get some put together for this. If not, I'll try to run the integration against a test account I have access to, to at least confirm the change works as expected.

BenB196 replied:
Following up here, I was able to test and determine that `EndDate` for all events is the same and takes its value from the `$filter` `EndDate`, so ordering here doesn't matter. However, while testing this, I observed the same issue with Microsoft's [24 hour eventual consistency](https://learn.microsoft.com/en-us/previous-versions/office/developer/o365-enterprise-developers/jj984335(v=office.15)#data-granularity-persistence-and-availability) and data gaps/missing data;

![image](https://github.com/elastic/integrations/assets/8277432/11a3f285-d9f4-4187-bee9-d36d19258758)

(Graph shows the integration running at an Interval of `1m`, and a "clear" `1m` gap between intervals.)

Looking into the fix for this, I suspect that [`Received`](https://learn.microsoft.com/en-us/previous-versions/office/developer/o365-enterprise-developers/jj984335(v=office.15)#fields) needs to be used somewhere within the `$filter` or `cursor`.

BenB196 replied:
I've pushed a change for the data gaps, see https://github.com/elastic/integrations/pull/8717#issuecomment-1859322741 for more details.

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

BenB196 commented:
I've updated the PR to include the ability to configure an additional look-back time. I chose `1h` as the default value here, somewhat arbitrarily, but during testing I found that this was generally "good enough" without hitting the negative side-effects of going with some value like `24h`.

I've updated the docs to include a mention this new field and the potential for missing events.

Unfortunately, Microsoft doesn't seem to offer a way to sort by a similar field like `event.ingested`, so need to leverage a raw look-back here and then dedupe events. (Deduping is already a part of this integration).

I've updated the main PR description with screenshots showing the difference between current and this PR.

I took the "wording" for this new field mainly from Kibana's Alert "Additional Look-back time" field.

ShourieG commented:
/test

ShourieG commented on packages/microsoft_exchange_online_message_trace/docs/README.md:
```diff
@@ -53,6 +53,11 @@ available in your organization. They are usually under the sections [Accepted Do
 - The default `Polling Interval` and `Initial Interval` values are configured to `1h`, you can however change these to your required values. The look-back 
   value of `Initial Interval` should not exceed `200 hours` as this might cause unexpected errors with the API.
 
+- The default `Additional Look-back Time` value is configured for to `1h`. 
```
```suggestion
- The default `Additional Look-back Time` value is configured for `1h`. 
```

ShourieG commented on packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:
```diff
@@ -53,6 +53,11 @@ available in your organization. They are usually under the sections [Accepted Do
 - The default `Polling Interval` and `Initial Interval` values are configured to `1h`, you can however change these to your required values. The look-back 
   value of `Initial Interval` should not exceed `200 hours` as this might cause unexpected errors with the API.
 
+- The default `Additional Look-back Time` value is configured for to `1h`. 
```
```suggestion
- The default `Additional Look-back Time` value is configured for `1h`. 
```

ShourieG commented on packages/microsoft_exchange_online_message_trace/_dev/build/docs/README.md:
```diff
@@ -53,6 +53,11 @@ available in your organization. They are usually under the sections [Accepted Do
   value of `Initial Interval` should not exceed `200 hours` as this might cause unexpected errors with the API.
 
+- The default `Additional Look-back Time` value is configured for to `1h`. 
+  This is intended to allow for catching events that were not initially present because of eventual consistency.
```
```suggestion
  This is intended to capture events that may not have been initially present due to eventual consistency.
```

elasticmachine commented:
### :rocket: Benchmarks report




To see the full report comment with `/test benchmark fullreport`
<!--COMMENT_GENERATED_WITH_ID_benchmark-report-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`1/1`) | :green_heart: 
Files | 100.0% (`1/1`) | :green_heart: 5.556
Classes | 100.0% (`1/1`) | :green_heart: 5.556
Methods | 100.0% (`15/15`) | :green_heart: 10.299
Lines | 98.106% (`259/264`) | :+1: 12.023
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

ShourieG commented:
@BenB196 The change looks good so far barring a few documentation issues, it would be awesome if you could add some system tests to simulate this updated behaviour of the integration.

BenB196 commented:
@ShourieG regarding the system tests. I'm a bit stuck here. I'd like to have a system test which queries the test endpoint, and asserts `2` (as it today). Then the test endpoint adds a 3rd entry and the agent queries it again and should assert `3`. The issue I'm having is I can't seem to figure out how to have the system `test-httpjson-config.yml` run two queries and assert two different values for each run.

Would you happen to know of a similar system test example or point to the docs which highlight how to implement a multi-run system test?

efd6 commented:
@BenB196 You can have 2 system test configs, one for the 2 doc case and one for the 3 doc case.

BenB196 commented:
> @BenB196 You can have 2 system test configs, one for the 2 doc case and one for the 3 doc case.

Hi @efd6 thanks for the reply. Question, will both tests then use the same Elastic Agent state config? The reason, I ask is I'd like the system test if possible; It is really one systems test that does both assertions.

Ideal test flow:

1. Spin up systems test
2. Publish 2 events
3. Elastic Agent Asserts `2`
4. Publish 3rd event between the initial 2 events
5. Elastic Re-reads event with existing agent state, and asserts `3` events.

This test would allow for covering both the `cursor` and `look-back` changes.

efd6 commented:
No, they would be separate tests. You would need to engineer the 2 and then 1 behaviour in the 3 event test. The 2 event test case would test the initial 2 events in the 3 event case.

BenB196 commented on packages/microsoft_exchange_online_message_trace/data_stream/log/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -10,6 +10,9 @@ processors:
       value: '8.11.0'
+  # Drop "null" event
+  - drop:
+      if: ctx.event.original == "{\"odata.metadata\":\"https://reports.office365.com/ecp/ReportingWebService/Reporting.svc/$metadata#MessageTrace\",\"value\":[]}"
```
Need to add this or else the tests fail with;

```
FAILURE DETAILS:
microsoft_exchange_online_message_trace/log httpjson:
[0] field "microsoft.online_message_trace.odata.metadata" is undefined


╭─────────────────────────────────────────┬─────────────┬───────────┬───────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──────────────╮
│ PACKAGE                                 │ DATA STREAM │ TEST TYPE │ TEST NAME │ RESULT                                                                                                                │ TIME ELAPSED │
├─────────────────────────────────────────┼─────────────┼───────────┼───────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────┤
│ microsoft_exchange_online_message_trace │ log         │ system    │ httpjson  │ FAIL: one or more errors found in documents stored in logs-microsoft_exchange_online_message_trace.log-ep data stream │  26.7124457s │
```

This comes from the trail/"break" value of; `.last_response.terminate_pagination`

efd6 replied:
The way that we normally deal with this kind of situation is to check whether `value` field is an array and is empty. For example elsewhere https://github.com/elastic/integrations/blob/e1354c1c5e511fd38834dbd8cc4950247a0ce74d/packages/infoblox_bloxone_ddi/data_stream/dns_config/elasticsearch/ingest_pipeline/default.yml#L21-L25. This is more costly than a string comparison in the failing case, but in the happy case, it's work that we already had to do and the test is more robust to changes in the API behaviour.

BenB196 replied:
Updated to check for a non-null `value` object after the json convert.

https://github.com/elastic/integrations/blob/7da5896aecfcbe2cf9eae1a80fc09526b172ea76/packages/microsoft_exchange_online_message_trace/data_stream/log/elasticsearch/ingest_pipeline/default.yml#L18-L19

BenB196 commented on packages/microsoft_exchange_online_message_trace/_dev/deploy/docker/files/config.yml:
```diff
@@ -34,7 +34,85 @@ rules:
+              {"Organization":"contoso.com","MessageId":"<a210cf91-4f2e-484c-8ada-3b27064ee5e3@az.uksouth.production.microsoft.com>","Received":"2022-09-05T18:10:13.4907658","SenderAddress":"azure-noreply@microsoft.com","RecipientAddress":"linus@contoso.com","Subject":"PIM: A privileged directory role was assigned outside of PIM","Status":"Delivered","ToIP":"216.160.83.56","FromIP":"81.2.69.144", ...
+            ]
+          }
+  - path: /ecp/reportingwebservice/reporting.svc/MessageTrace
+    methods: [ GET ]
+    request_headers:
+      Authorization:
+        - "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ilg1ZVhrN"
+    query_params:
+      $top: "{$top:1}"
+      $skiptoken: "{$skiptoken:2}"
+    responses:
+      - status_code: 200
+        headers:
+          Content-Type:
+            - "application/json"
+        body: |-
+          {
+            "odata.metadata":"https://reports.office365.com/ecp/ReportingWebService/Reporting.svc/$metadata#MessageTrace",
+            "value": []
+          }
```
This adds an actual test for validating the trailing of an batch, and allows for using `interval` within the system test.

efd6 replied:
It would be good to add the pipeline test for this too. The test corresponding to the example above is here https://github.com/elastic/integrations/blob/e1354c1c5e511fd38834dbd8cc4950247a0ce74d/packages/infoblox_bloxone_ddi/data_stream/dns_config/_dev/test/pipeline/test-pipeline-dns-config.log#L4

BenB196 replied:
Pipeline test added;

https://github.com/elastic/integrations/blob/7da5896aecfcbe2cf9eae1a80fc09526b172ea76/packages/microsoft_exchange_online_message_trace/data_stream/log/_dev/test/pipeline/test-log.log#L5
https://github.com/elastic/integrations/blob/7da5896aecfcbe2cf9eae1a80fc09526b172ea76/packages/microsoft_exchange_online_message_trace/data_stream/log/_dev/test/pipeline/test-log.log-expected.json#L440

BenB196 reviewed (commented):
(This comment is partially outdated, see: https://github.com/elastic/integrations/pull/8717#discussion_r1432183340)

I've expanded on the system tests here a bit. 

There is still an issue with this though that I'll need to take a look at tomorrow, but pushing in the event someone else wants to take a look.

The current issue is that the `cursor`

```yaml
cursor:
  last_execution_datetime:
    value: '[[ toJSON .last_event.message "EndDate" ]]'
```

Can become null/empty, in the event that the `.last_response` is empty, more specifically, you get `value: []` (which it can be), and defaults back to `now`. I'm not sure the best way to handle this issue currently.

A few options I see:

1. Default to `now`.
    - This is the easiest, but has the disadvantage of bringing back some of the issues this PR initially tried to solve.
2. Find a way to parse the `$filter` param to try and get the EndDate value from it.
    - This would provide the current value, but I don't see a way to easily parse this param to get the value within the httpjson templating.
3. Open to ideas/suggestions that people might have.

efd6 commented:
/test

BenB196 commented on packages/microsoft_exchange_online_message_trace/data_stream/log/_dev/test/system/test-httpjson-null-trailing-config.yml:
Added a systems test to cover the null/empty trailing event case.

BenB196 commented on packages/microsoft_exchange_online_message_trace/data_stream/log/agent/stream/httpjson.yml.hbs:
```diff
@@ -66,11 +66,11 @@ response.pagination:
 cursor:
   last_execution_datetime:
-    value: '[[(formatDate (now) "RFC3339")]]'
+    value: '[[ slice (.last_response.url.params.Get "$filter") 68 88 ]]'
```
This works by looking the the `$filter` param, and slicing/splitting to only the value of the `EndDate` time.

Example:

`StartDate+eq+datetime'2023-12-19T22:47:25Z'+and+EndDate+eq+datetime'2023-12-19T23:47:25Z'`

Gets parsed down to

`2023-12-19T23:47:25Z`

Note, unfortunately I don't think it is possible to add a system test for this as the testing tools stand today. The issue:

- This integration heavily relies on `now` for values, making idempotent system tests impossible (from what I can tell)
- [stream](https://github.com/elastic/stream) does not support dynamic query parameters, so you can't really scope a rule in the test to be hit after `X` duration of time since the test started or something, which you'd need to do for dynamic values of `StartDate` or `EndDate`.

I'm not sure if I'm missing something or if there are any ideas on how to do this via a system test.

You can manually test & validate this though. If you run the `httpjson` system test, and look at the log file for `exchange_online`. You should see properly incrementing `StartDate` and `EndDate` values.

Example:

```
# Interval 0
$filter=StartDate+eq+datetime'2023-12-20T01:02:48Z'+and+EndDate+eq+datetime'2023-12-20T02:02:48Z'
$filter=StartDate+eq+datetime'2023-12-20T01:02:48Z'+and+EndDate+eq+datetime'2023-12-20T02:02:48Z'
# Interval 1
$filter=StartDate+eq+datetime'2023-12-20T01:02:48Z'+and+EndDate+eq+datetime'2023-12-20T02:02:53Z'
$filter=StartDate+eq+datetime'2023-12-20T01:02:48Z'+and+EndDate+eq+datetime'2023-12-20T02:02:53Z'
# Interval 2
$filter=StartDate+eq+datetime'2023-12-20T01:02:53Z'+and+EndDate+eq+datetime'2023-12-20T02:02:58Z'
$filter=StartDate+eq+datetime'2023-12-20T01:02:53Z'+and+EndDate+eq+datetime'2023-12-20T02:02:58Z'
```

You can see that the `StartDate` of the next interval is the `EndDate` of the previous interval.

I have also tested this integration now against an test O365 environment and confirmed that it is working as expected.

ShourieG replied:
This seems it will work as long as the filter query does not change its character size. Do we know that this works well for something like top=1000,skiptoken=10000, etc ? cause I'm guessing larger values of batch size will drive these up increasing the char length of the filter query ... In this scenario the character positions provided for size will become invali.

ShourieG replied:
@BenB196 This would only work if the length of the $filter query remains same forever, do we know for sure that the startDate and and endDate character count will always be [ XXXX-XX-XXXXX:XX:XXX ] ? Also this would make further additions to filter in future (if required) trickier as one would need to re-evaluate the slice positions. What do you think @efd6, I'm personally a bit concerned about this approach ?

efd6 replied:
They don't specify a specific format within ISO8601 that the start date and end date are rendered as, so I would be extremely reluctant to rely on this.

ShourieG replied:
Yea exactly, this is basically a limitation of the api response, which does not have updated endDate that's causing this issue in the first place, if we cannot guarantee that this approach will always work as expected then we should step back and think on what else could be done here. Maybe 
```js
cursor:
  last_execution_datetime:
    value: '[[ toJSON .last_event.message "EndDate" ]]' 
```
is the only possible approach for now even if it brings back some inconsistency from before. The only other options I see here is to -

1) Change the behaviour when the cursor value is actually calculated (end of op) - highly risky as it will impact multiple existing integrations.
2)  Move to CEL, as CEL has various mechanism for dealing with similar issues.
3) Introduce something new in httpjson template options to deal with parsing out such values.

BenB196 replied:
> This would only work if the length of the $filter query remains constant, do we know for sure that the startDate and and endDate character count will always be [ XXXX-XX-XXXXX:XX:XXX ] ? Also this would make further additions to filter in future (if required) trickier as one would need to re-evaluate the slice positions.

We know for sure that the dates will always be the same length, as we always format them with Go's `RFC3339` implementation.

https://github.com/elastic/integrations/blob/b7b7501859eab8b40aa20b08227994f1ffbe243b/packages/microsoft_exchange_online_message_trace/data_stream/log/agent/stream/httpjson.yml.hbs#L31-L32

Whether someone adds an additional filter is a bit harder. There isn't a way to "unit test" with integrations, so in theory someone could add another filter value, but based on the current Microsoft Reporting API and this package use case I don't see this happening in the near future.

```yaml
cursor:
  last_execution_datetime:
    value: '[[ toJSON .last_event.message "EndDate" ]]' 
```

^ This turns out doesn't actually work. From further testing I found that it would always cause a templating error and cause the request transform to go back to `default`

https://github.com/elastic/integrations/blob/b7b7501859eab8b40aa20b08227994f1ffbe243b/packages/microsoft_exchange_online_message_trace/data_stream/log/agent/stream/httpjson.yml.hbs#L32C1-L32C1

Making it look like it was working, but not actually working.

---

While

```yaml
cursor:
  last_execution_datetime:
    value: '[[ slice (.last_response.url.params.Get "$filter") 68 88 ]]'
```

Does look brittle (and is to a degree), I think it might be the "best" solution that is currently available for this integration as it stands using httpjson.

I do agree that this integration probably needs one of these:

> 1. Change the behaviour when the cursor value is actually calculated (end of op) - highly risky as it will impact multiple existing integrations.
> 2. Move to CEL, as CEL has various mechanisms for dealing with similar issues.
> 3. Introduce something new in httpjson template options to deal with parsing out such values.

But I also think that there probably should be an interim "fix" of some kind here. As it stands the current 1.17.x version only seems to be collecting ~60-80% of events. This PR currently brings that up to 95-98%, so while not great and still a bit fragile, does improve the current standing.

efd6 replied:
I think you have convinced me. It does need a comment justifying it so that when it does break, we have a clear hint about why.

BenB196 replied:
Pushed https://github.com/elastic/integrations/pull/8717/commits/e6cd2566f7196cf0da43f642984053451569b1eb to add comments warning about `$filter` modification.

efd6 commented on packages/microsoft_exchange_online_message_trace/data_stream/log/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -14,6 +14,9 @@ processors:
       target_field: microsoft.online_message_trace
+  # Drop "null" event
+  - drop:
+      if: ctx.microsoft?.online_message_trace?.value != null
```
Let's make the test more specific.
```suggestion
      if: ctx.microsoft?.online_message_trace?.value instanceof List && ctx.microsoft.online_message_trace.value.length == 0 
```

efd6 commented:
/test

efd6 commented on packages/microsoft_exchange_online_message_trace/data_stream/log/agent/stream/httpjson.yml.hbs:
```diff
@@ -66,11 +67,14 @@ response.pagination:
 cursor:
   last_execution_datetime:
-    value: '[[(formatDate (now) "RFC3339")]]'
+    # Expects `$filter` param to be in the format of: StartDate+eq+datetime'2023-12-19T22:47:25Z'+and+EndDate+eq+datetime'2023-12-19T23:47:25Z'
+    # Produces output of: 2023-12-19T23:47:25Z
+    # If you modify `$filter` format you might need to update the slice range
```
For complete avoidance of ambiguity.

```suggestion
    # Expects `$filter` param to be in the format of:
    #
    #  StartDate+eq+datetime'2023-12-19T22:47:25Z'+and+EndDate+eq+datetime'2023-12-19T23:47:25Z'
    #                                                                      ^~~~~~~~~~~~~~~~~~~~
    #
    # Produces output of: 2023-12-19T23:47:25Z
    # If you modify `$filter` format you might need to update the slice range
```

efd6 commented:
/test

efd6 reviewed (approved):
LGTM, but waiting for @ShourieG.

ShourieG reviewed (approved):
LGTM

elasticmachine commented:
Package microsoft_exchange_online_message_trace - 1.18.0 containing this change is available at https://epr.elastic.co/search?package=microsoft_exchange_online_message_trace

elasticmachine commented:
Package microsoft_exchange_online_message_trace - 1.18.0 containing this change is available at https://epr.elastic.co/search?package=microsoft_exchange_online_message_trace
